o/snapstate: use refresh hints to continue auto-refresh after monitoring

MiguelPires opened the pull request:
This addresses the [issue of Firefox auto-refreshes not being continued](https://bugs.launchpad.net/snapd/+bug/2019158) after the user stops the snap. The root cause is that the second auto-refresh doesn't always get a consistent response from the store. That is, the first auto-refresh might get a new revision but, if it's inhibited by a running snap, the next auto-refresh might not. This only happens for Firefox (to my knowledge) because the store only throttles Firefox, to reduce the load on its servers. This PR uses the refresh hints written during the auto-refresh (added [here](https://github.com/snapcore/snapd/pull/12946)) to continue the auto-refresh without relying on the store to return the same information.

pedronis commented on overlord/snapstate/handlers.go:
```diff
@@ -834,16 +834,10 @@ func (m *SnapManager) doPreDownloadSnap(t *state.Task, tomb *tomb.Tomb) error {
 	}
 
 	targetFn := snapsup.MountFile()
-	dlOpts := &store.DownloadOptions{
-		// pre-downloads are only triggered in auto-refreshes
-		Scheduled: true,
-		RateLimit: autoRefreshRateLimited(st),
-	}
-
```
also not sure why this change?

MiguelPires replied:
I think I removed it when the field was AutoRefresh, I'm not sure the scheduled header makes sense for the pre-download. In particular, since we now know that in other cases it's used to omit a change (which I assume would never apply to a download but still). Do you think I should leave it?

pedronis replied:
it is still part of an auto-refresh, also we for sure want the ratelimit

pedronis commented on overlord/snapstate/handlers.go:
```diff
@@ -892,113 +886,196 @@ func (m *SnapManager) doPreDownloadSnap(t *state.Task, tomb *tomb.Tomb) error {
 	// notify the user about the blocked refresh
 	asyncPendingRefreshNotification(context.TODO(), userclient.New(), refreshInfo)
 
+	var refreshHints map[string]*refreshCandidate
+	if err := st.Get("refresh-candidates", &refreshHints); err != nil {
+		return fmt.Errorf("cannot get refresh-candidates: %v", err)
+	}
```
refreshHints doesn't seem to be used in this function itself?

MiguelPires replied:
Ah it's a leftover from when it got moved into the function below. I'll remove it

pedronis commented on overlord/snapstate/handlers.go:
```diff
@@ -892,113 +886,196 @@ func (m *SnapManager) doPreDownloadSnap(t *state.Task, tomb *tomb.Tomb) error {
-	monitoredSnaps[refreshInfo.InstanceName] = abort
-	updateMonitoringState(st, monitoredSnaps)
+	if err := addMonitoring(st, snapName, abort); err != nil {
+		return fmt.Errorf("cannot save monitoring state: %v", err)
```
should this be an error or only something that we log?

MiguelPires replied:
I think it makes sense to bubble up the error and fail the pre-download change since we're not sure whether we saved an abort channel or monitoring info. If we keep going, then we might not have state to recover from restarts, abort the monitoring or determine whether there's a monitoring goroutine already. So that might result in weird scenarios and bugs

pedronis replied:
I meant log and return, not log and continue. But maybe that was clear?

MiguelPires replied:
Ah sure, yeah I can change that

MiguelPires commented on overlord/snapstate/snapstate.go:
```diff
@@ -1707,7 +1707,7 @@ func updateManyFiltered(ctx context.Context, st *state.State, names []string, re
 	names = strutil.Deduplicate(names)
 
-	refreshOpts := &store.RefreshOptions{Scheduled: flags.IsAutoRefresh}
+	refreshOpts := &store.RefreshOptions{Scheduled: flags.IsAutoRefresh && !flags.IsContinuedAutoRefresh}
```
True, I'll remove this

MiguelPires commented on overlord/snapstate/handlers.go:
```diff
@@ -736,7 +736,7 @@ func (m *SnapManager) doDownloadSnap(t *state.Task, tomb *tomb.Tomb) error {
 	perfTimings := state.TimingsForTask(t)
 	snapsup, theStore, user, err := downloadSnapParams(st, t)
-	if snapsup != nil && snapsup.IsAutoRefresh {
+	if snapsup != nil && snapsup.IsAutoRefresh && !snapsup.IsContinuedAutoRefresh {
```
Removed

MiguelPires commented on overlord/snapstate/handlers.go:
```diff
@@ -753,7 +753,7 @@ func (m *SnapManager) doDownloadSnap(t *state.Task, tomb *tomb.Tomb) error {
 
 	dlOpts := &store.DownloadOptions{
-		Scheduled: snapsup.IsAutoRefresh,
+		Scheduled: snapsup.IsAutoRefresh && !snapsup.IsContinuedAutoRefresh,
```
Removed

pedronis reviewed (commented):
thanks for the changes, question about what should happen when a snap disappears from refresh-candidates. Also should we check if changing refresh-candidates drop snaps, and should we stop their monitoring there?

MiguelPires commented on overlord/snapstate/autorefresh.go:
```diff
@@ -98,16 +101,23 @@ func (rc *refreshCandidate) InstanceName() string {
-	return &rc.SnapSetup, &snapst, nil
+
+	snapsup := &rc.SnapSetup
+	if globalFlags != nil {
+		snapsup.Flags.IsAutoRefresh = globalFlags.IsAutoRefresh
+		snapsup.Flags.IsContinuedAutoRefresh = globalFlags.IsContinuedAutoRefresh
+	}
```
Added a test for this

MiguelPires commented:
> Also should we check if changing refresh-candidates drop snaps, and should we stop their monitoring there?
 
I added some code to the `refreshhints.go` that aborts the monitoring routines of snaps for which hints were dropped from refresh-candidates (if that happens because something was already refreshed, the monitoring was already aborted so it's a no-op). 

I also changed the abort channel buffer to be larger than 1 because we can now have several writers to it. The state locks make it very unlikely but I think it's possible for a ``--ignore-running` refresh to cause a write to the abort channel, followed by `refreshhints.go` getting the state lock before the reader, writing to the channel again and deadlocking. I was thinking about possible tests for this but it seems quite difficult since it depends the timing of different handlers and goroutines getting the state lock.

pedronis commented on overlord/snapstate/handlers.go:
```diff
@@ -948,14 +961,18 @@ func addMonitoring(st *state.State, snapName string, abort chan<- bool) error {
+
 		// should never happen save for programmer error
-		return fmt.Errorf(`"monitored-snaps" should be map[string]chan<- bool but got %T`, storedChans)
+		return false, fmt.Errorf(`"monitored-snaps" should be map[string]chan<- bool but got %T`, storedChans)
```
this should be marked internal error I suppose

pedronis replied:
I shared more code around this and addressed this that way

pedronis commented on overlord/snapstate/handlers.go:
```diff
@@ -915,7 +915,10 @@ func asyncRefreshOnSnapClose(st *state.State, snapsup *SnapSetup, refreshInfo *u
+	// use a buffer larger than 1 because there can be more than one writes to the channel.
+	// State locks make it unlikely but it's possible to have a second write before
+	// the reader can drain and remove the channel from the state
+	abort := make(chan bool, 16)
```
we should probably switch to a cancellable context instead

pedronis commented on overlord/snapstate/refreshhints.go:
```diff
@@ -97,10 +97,35 @@ func (r *refreshHints) refresh() error {
 		return fmt.Errorf("internal error: cannot get refresh-candidates: %v", err)
 	}
+
+	stopMonitoringOutdatedCandidates(r.state, hints)
+
 	r.state.Set("refresh-candidates", hints)
```
do we need this logic in other places that set refresh-candidates ?

pedronis replied:
created a helper for this to use consistently

pedronis reviewed (commented):
thanks for the latest set of changes, some comments, I'll try to take care of them

ernestl reviewed (approved):
LGTM
